Source/DafnyCore/Resolver/PreType/PreTypeToType.cs

```diff
@@ -0,0 +1,174 @@
+      default:
+        break;
+    }
+    var arguments = pt.Arguments.ConvertAll(PreType2Type);
+    if (pt.Decl is ValuetypeDecl valuetypeDecl) {
+      return valuetypeDecl.CreateType(arguments);
+    } else if (pt.Decl is ArrowTypeDecl arrowTypeDecl) {
+      return new ArrowType(pt.Decl.tok, arrowTypeDecl, arguments);
+    } else {
+      return new UserDefinedType(pt.Decl.tok, pt.Decl.Name, pt.Decl, arguments);
+    }
+  }
+
+  protected override void VisitOneDeclaration(TopLevelDecl decl) {
+    if (decl is NewtypeDecl newtypeDecl) {
+      UpdateIfOmitted(newtypeDecl.BaseType, newtypeDecl.BasePreType);
+    } else if (decl is SubsetTypeDecl subsetTypeDecl) {
+      UpdateIfOmitted(subsetTypeDecl.Var.Type, subsetTypeDecl.Var.PreType);
+    }
+    base.VisitOneDeclaration(decl);
+  }
+
+  private void UpdateIfOmitted(Type type, PreType preType) {
+    if (type is TypeProxy { T: null } typeProxy) {
+      typeProxy.T = PreType2Type(preType);
+    }
+  }
+
+  private void UpdateTypeOfVariables(IEnumerable<IVariable> variables) {
+    foreach (var v in variables) {
+      UpdateIfOmitted(v.Type, v.PreType);
+    }
+  }
+
+  public override void VisitField(Field field) {
+    // The type of the const might have been omitted in the program text and then inferred
+    UpdateIfOmitted(field.Type, field.PreType);
+    base.VisitField(field);
+  }
+
+  protected override void PostVisitOneExpression(Expression expr, IASTVisitorContext context) {
+    if (expr is FunctionCallExpr functionCallExpr) {
+      functionCallExpr.TypeApplication_AtEnclosingClass = functionCallExpr.PreTypeApplication_AtEnclosingClass.ConvertAll(PreType2Type);
+      functionCallExpr.TypeApplication_JustFunction = functionCallExpr.PreTypeApplication_JustFunction.ConvertAll(PreType2Type);
+    } else if (expr is MemberSelectExpr memberSelectExpr) {
+      memberSelectExpr.TypeApplication_AtEnclosingClass = memberSelectExpr.PreTypeApplication_AtEnclosingClass.ConvertAll(PreType2Type);
+      memberSelectExpr.TypeApplication_JustMember = memberSelectExpr.PreTypeApplication_JustMember.ConvertAll(PreType2Type);
+    } else if (expr is ComprehensionExpr comprehensionExpr) {
+      UpdateTypeOfVariables(comprehensionExpr.BoundVars);
+    } else if (expr is LetExpr letExpr) {
+      UpdateTypeOfVariables(letExpr.BoundVars);
+      foreach (var lhs in letExpr.LHSs) {
+        VisitPattern(lhs, context);
+      }
+    } else if (expr is DatatypeValue datatypeValue) {
+      Contract.Assert(datatypeValue.InferredTypeArgs.Count == 0 || datatypeValue.InferredTypeArgs.Count == datatypeValue.InferredPreTypeArgs.Count);
+      if (datatypeValue.InferredTypeArgs.Count == 0) {
+        foreach (var preTypeArgument in datatypeValue.InferredPreTypeArgs) {
+          datatypeValue.InferredTypeArgs.Add(PreType2Type(preTypeArgument));
+        }
+      }
+    }
+
+    if (expr.PreType is UnusedPreType) {
+      expr.Type = new InferredTypeProxy();
```

Comment:
Perhaps my biggest concern is related to this: is the benefit of not having to annotate a constant with its type worth the cost? Is it even beneficial? We require the users to annotate their functions with types in part because it makes the code more readable, why not constants as well?

Reply:
I hear your concern. Indeed, maybe the language should be changed to insist that every `const` (with or without a RHS) is given an explicit type, and that the base type of `newtype` and subset types are given explicitly. In my next PR, I handle the inference of these things, so I suggest we delay the decision of whether or not to change the language.